### ISSUE TITLE
Tracking Issue for Provider API

### ISSUE THREAD

**Issue by yaahc** [rust-lang/rust#96024]
Feature gate: `#![feature(provide_any)]`

This is a tracking issue for [RFC 3192](https://github.com/rust-lang/rfcs/pull/3192) 

> This RFC proposes extending the [`any` module](https://doc.rust-lang.org/nightly/std/any/index.html) of the core library with a generic API for objects to provide type-based access to data. (In contrast to the existing APIs which provides type-driven downcasting, the proposed extension integrates downcasting into data access to provide a safer and more ergonomic API).

The initial motivating use case for this API is to provide the necessary functionality to add [generic member access](https://github.com/rust-lang/rfcs/pull/2895) to the `Error` trait to generalize extracting context like `Backtrace` and to make it possible to move the `Error` trait into `core` while leaving `Backtrace` in `std`.

### Public API

```rust
// core::any
pub trait Provider {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>);
}

pub fn request_value<'a, T>(provider: &'a (impl Provider + ?Sized)) -> Option<T>
where
    T: 'static;

pub fn request_ref<'a, T>(provider: &'a (impl Provider + ?Sized)) -> Option<&'a T>
where
    T: 'static + ?Sized;

#[repr(transparent)]
pub struct Demand<'a>(_);

impl<'a> Demand<'a> {
    pub fn provide_value<T>(&mut self, value: T) -> &mut Self
    where
        T: 'static;

    pub fn provide_value_with<T>(&mut self, fulfil: impl FnOnce() -> T) -> &mut Self
    where
        T: 'static;

    pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self;

    pub fn provide_ref_with<T: ?Sized + 'static>(
        &mut self,
        fulfil: impl FnOnce() -> &'a T,
    ) -> &mut Self;

    pub fn would_be_satisfied_by_value_of<T>(&self) -> bool
    where
        T: 'static;

    pub fn would_be_satisfied_by_ref_of<T>(&self) -> bool
    where
        T: ?Sized + 'static,
}

impl<'a> fmt::Debug for Demand<'a> {}

// core::error
impl<E> Provider for E
where
    E: Error + ?Sized;
```

### Steps / History

- [x] Implementation: https://github.com/rust-lang/rust/pull/91970
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by yaahc** [comment]
https://github.com/rust-lang/rust/pull/98912 updates the `request_*` functions to have an APIT (arg position impl trait) parameter instead of a standard generic parameter to remove unnecessary inferred parameters while turbofishing.

Callers change:

```rust
let obj: Box<dyn Provider> = Box::new(SomeConcreteType { some_string: "hello".to_owned() });

// From
assert_eq!(&**request_ref::<String, _>(&*obj).unwrap(), "hello");
assert_eq!(&*request_value::<String, _>(&*obj).unwrap(), "bye");
assert_eq!(request_value::<u8, _>(&*obj), None);

// To
assert_eq!(&**request_ref::<String>(&*obj).unwrap(), "hello");
assert_eq!(&*request_value::<String>(&*obj).unwrap(), "bye");
assert_eq!(request_value::<u8>(&*obj), None);
```

**Comment by dtolnay** [comment]
One minor thing that weirds me out about this API is this impl:

https://github.com/rust-lang/rust/blob/02654a0844f5c8d29bac318c3c6c666da3d8543d/library/core/src/error.rs#L197-L199

Since this is a blanket impl (`Self` type is a generic parameter), it is the _one_ blanket impl that `Provider` is ever going to be able to have, at least absent some far-future negative impls feature in the language. Any other blanket impl would be considered overlapping and violate coherence.

But it's a blanket impl of `core::any::Provider`, which is envisioned as being as general-purpose as the `Any` trait, for types which implement `core::error::Error`, which is a much narrower use case. In some sense this relationship is saying that "Errors are the one true canonical/privileged implementers of the Provider trait; any other application of Provider is second-class".

I am wondering which one of the following would be feasible:

1. Yes, errors are the one privileged application of Provider, and while Provider may be useful for some other things sometimes, all of those are second-class.

    a. Maybe move Provider to `core::error` to double down on this view?

2. Remove the blanket impl. What are the things that make the blanket impl necessary for usability of this API, and can those be mitigated a different way?

**Comment by yaahc** [comment]
We could go back to what we had before we move Error into core which was a provide impl just on dyn Error which caused some other paper cuts but I can't remember exactly what those issues were.

**Comment by thomcc** [comment]
Another issue for the provider API that matters for non-error cases is the lack of support for &mut.

I think the initial proposal did support this, at some cost to complexity. While I don't know that we need to support it out of the gate (I've wanted it multiple times when playing around with the API), it would be good to be sure we know it would look like, so that if we want to add it in the future, that we have not painted ourselves into a corner where we cannot do so.

**Comment by yaahc** [comment]
@thomcc I believe all we have to do is expose the tag APIs to support those cases. We didn't change the impl, only the exposed API.

**Comment by JonasJebing** [comment]
> We could go back to what we had before we move Error into core which was a provide impl just on dyn Error which caused some other paper cuts but I can't remember exactly what those issues were.

Here's the discussion that lead to `impl<E: Error + ?Sized> Provider for E`: https://github.com/rust-lang/rust/pull/98072#discussion_r898313181
So, the paper cut without the blanket impl is, that it's not possible to call `core::any::request_*` or `<dyn Error>::request_*` on a concrete error type like `std::io::Error`.

Personally, I feel like there are a lot of possible workarounds for that paper cut:
1. The concrete error type could allow public access to its fields or getter functions;
2. The concrete error type could `impl Provider for SomeConcreteError`
3. A user can always create an adapter/newtype that implements Provider using `Error::provide` on the conrete error type.
So, I am leaning towards `impl Provider for dyn Error` instead of the blanket impl.

**Comment by JonasJebing** [comment]
@jakobhellermann you're missing that `Provider::provide` just has a shared reference `&self`.
So, a `fn provide_mut<'a>(&'a mut self, _req: &mut Demand<'a>) {}` needs to be added to the Provider trait, too.

@yaahc, @thomcc the Provider support for &mut was addressed in the [RFC under "Future Possibilities"](https://rust-lang.github.io/rfcs/3192-dyno.html#future-possibilities).
The linked repository even includes an API for providing different &mut types to the same field or struct:
~~~rust
    fn provide_mut<'a>(&'a mut self, req: &mut Demand<'a>) {
        req.with(self)
            .provide_mut::<String>(|this| &mut this.s)
            .provide_mut::<Bar>(|this| this);
    }
~~~

**Comment by Kixunil** [comment]
First of all, I think this idea is really great so big thanks to all people participating on the design and implementation!

One thing that's puzzling me is the limitation of `T: 'static`. I don't see any reason for it besides possible compatibility with `Any`. (Could there be simply another method enforcing `static`?)
What I'd like to do is to provide singular value that may reference data inside the error. To be specific I'm very interested in having a common API for spanned errors that can be displayed nicely while being detached from rendering (`codespan-reporting`, `miette`...)

```rust
pub struct SpannedLabel<'a> {
    span: Span,
    label: &'a dyn fmt::Display, // maybe this should be an enum supporting `String` and `Box<fmt::Display>`
    file_id: FileId, // didn't figure out the exact type yet
}

#[derive(Copy, Clone /* ... */)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}
```

One way to resolve this issue is to provide the values separately. This is not entirely unreasonable but I worry it'll lead to people forgetting about providing one of the two values.

Also I found this in the motivation section of the RFC:

> this approach is limited by constraints on `Any`, we cannot return objects by value, return objects which include references (due to the `'static` bound on `Any`),

But this still only works with `'static` and is even acknowledged by the RFC:

> The proposal only supports handling value types (or other types with only `'static` lifetimes) and reference types.

Which sounds contradictory and I don't see it explained anywhere.

**Comment by conradludgate** [comment]
> One thing that's puzzling me is the limitation of T: 'static

I'm not sure on the nitty gritty details of `typeid` and `Any` (and I hear there's a lot of cursed stuff to know) but  `T: 'a` and `T: 'b` would definitely need to be different types (need to be very careful with subtyping etc).

Now, playing devils advocate as a non compiler engineer, I think trying to make sure that `T: 'a` and `T: 'b` have different unique typeids would be a lot of work, especially since we have troubles with that for `T` and `U` anyway

**Comment by Kixunil** [comment]
@conradludgate I don't see how it's relevant. AFAIK this implementation bypasses `typeid` and `Any` completely.

**Comment by conradludgate** [comment]
> AFAIK this implementation bypasses typeid and Any completely.

It bypasses `Any` (using a trait called `Erased` instead), but still uses `TypeId`

https://github.com/rust-lang/rust/blob/8e6809072304b147f9e98d55c87f42c0f3959679/library/core/src/any.rs#L1285-L1287

**Comment by mstone** [comment]
(With apologies if this is a bad place to ask:) while we are discussing continuing appearances of this "`type_id()` is lifetime-insensitive" limitation w.r.t. to type-directed programming + reflection, is there additional history or context, such as a good issue to link to, for those who would like to know more about the overall problem?

(Also, if it helps, what I've got for context so far mostly consists of, going from older to more recent:

* https://rust-lang.github.io/rfcs/1849-non-static-type-id.html
* https://github.com/rust-lang/rust/issues/41875
* https://internals.rust-lang.org/t/pre-rfc-non-footgun-non-static-typeid/17079

as well as for the sub-issue that arises within some of these conversations of not breaking inference while adding generics (with defaults where necessary) to existing signatures:

* https://faultlore.com/blah/defaults-affect-inference/ )

**Comment by arniu** [comment]
Rename the trait `Provider` to `Provide`?

**Comment by thomcc** [comment]
For various work-related reasons I had to do an audit of all of the current I-unsound issues in Rust a few months ago. I came away from it with the perhaps somewhat pessimistic conclusion where I start to mistrust any code with multi-indirect lifetimes on trait objects and function pointers (usually where contravariance and implied bounds are involved). This is https://github.com/rust-lang/rust/issues/25860, https://github.com/rust-lang/rust/issues/84366, https://github.com/rust-lang/rust/issues/100051, and a good handful of others.

The patterns used here, while they *should* be fine, have started to worry me, since they're awfully similar to the patterns that end up showing up in code that manages to find a gap in our lifetime checking. And the fact that some parts of this are a way to avoid the `'static` requirement which is normally present on `Any`... well, it doesn't make me feel that much better about it. That said I have no clue if this is actually the case or not -- this is 100% a vibes-based comment[^1].

Has (or could) someone with more specific knowledge into the way our current lifetime checking is faulty make sure this code isn't walk directly into those issues? (And that the API / code using it won't be broken when those issues are fixed?)

CC @steffahn, @lcnr... @rust-lang/types, I suppose?

[^1]: And I wasn't going to make a comment like this since it's basically just paranoia, but @scottmcm suggested I do so, so blame them.

**Comment by joshtriplett** [comment]
We had a *very long* conversation about the Provider API in a libs meetup today.

We came to a few conclusions:

Provider is really complicated. :)

We agreed with [@dtolnay's point](https://github.com/rust-lang/rust/issues/96024#issuecomment-1232310693) that there shouldn't be a blanket impl of `Provider` for all `Error`; that makes `Provider` much less useful as a generic type.

It seems like much of the complexity here is going into trying to make `Error` directly support `Provider` *without* making users of the provider API require `Provider` in addition to `Error`.

It's worth asking *if* we could just require things that want to ask for attached items via the provider API to require the `Provider` trait explicitly (not just `Error`).

That said, if we're going to keep that property and keep something approximating the current interface, one possible approach would be to drop the blanket impl, and have two `request_ref` functions: have `core::any::request_ref` take `impl Provider`, and have `core::error::request_ref` take `impl Error`. In both cases, the `request_ref` implementations could share their implementation, by having an internal helper that takes `thing: &T, provide_func: impl FnOnce(&T, &mut Demand)`, and then the `request_ref` impls become trivial calls to that helper.

We would not necessarily need to stabilize the general `Provider` trait right away, or for that matter the helper. We'd need to stabilize the method on `Error`, and the free functions.

**Comment by joshtriplett** [comment]
@the8472 also has a possible approach that would integrate this into the compiler and make it work on any type, via vtables. I feel that that would add substantial complexity, but it *would* have more capabilities. I will leave it to @the8472 to summarize and make the case for that approach.

**Comment by the8472** [comment]
My rough idea can be phrased as as opt-in sidecasting or a static field-offset/getter function lookup table. Either way the set of provided properties would have to be statically defined could then be accessed with a type-based lookup.

On the assembly level it would roughly (variations are possible) look like this:

- the vtable contains a pointer to a secondary table. this pointer can be null
- this new table contains key-value pairs consisting of a type ID as key and a pointer or usize
- there would be a builtin function that takes a reference, a type-id and performs the provide-op


Variations:

1.) the most basic one is to limit it to things that have a vtable (dyn types) and only field offsets and always return references. in this case the indirection is not necessary and this could be part of the vtable itself
2.) a) a possible extension is to allow the pointers in the new table to be function pointers. this way for each provided type a separate getter could be called
2.) b) alternatively each entry could be a vtable. this way side-casting from one dyn-trait to another one would be enabled. Types would have to opt into sidecasting to specific other types, this limits what tables and entries have to be generated
3.) the size and align entries of each vtable could be moved to the new table and shared across types of similar size, alignment and with an empty set of provided types (binary size wins!)

**Comment by the8472** [comment]
I got this idea because I don't like how the central `provide()` method essentially leads implementers to a "spray and pray" approach of just providing any values that might be requested.
This happens at runtime. It's intransparent to tools. It's like we designed a Rust feature on Opposite Day.

Adapter from the [module documentation](https://doc.rust-lang.org/std/any/index.html#examples-1) (which only has a single call):

```
impl Provider for SomeConcreteType {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
        demand.provide_ref::<String>(&self.some_string);
        demand.provide_ref::<Backtrace>(&self.bt);
        demand.provide_ref::<HttpErrorCode>(&self.code);
        demand.provide_ref::<TraceId>(&self.trace);
        // ....
    }
}
```

As can be seen, this just blindly everything this could be requested. For cheap field accesses this might be ok. But if such a method would have to do any expensive lookups (via global state or interior mutability) this could become expensive even if some particular thing isn't requested. Granted `would_be_satisfied_by_value_of` has been added but I'm not sure if this will always be used, especially if this becomes macro-generated.

**Comment by shepmaster** [comment]
@the8472 one usage I have for the provider API in SNAFU is the ability to "chain" provided values to causal errors. Ignoring exact syntax...

- Error type `Inner` provides a `Backtrace`
- Error type `Outer` wraps an `Inner`
- `Outer` can provide a `Backtrace` by delegating to `Inner`

Notably, `Outer` doesn't statically declare that it will provide a `Backtrace` — it's up to the underlying error. Additionally, it's not based on the *type* of the underlying error but the *value*. For example, the inner error may be an enum with two variants, one with and one without a backtrace.

Would your proposed change handle this dynamic behavior?

That being said, making more stuff known statically would change the calculus on how these provided data will be treated from a semver perspective, which is something I brought up during initial development of the feature. However, it could make things much more strict than many people might expect. Personal experience has shown that too many people don't think of error types as "real" types and thus don't worry about making breaking changes to error types.

---

My more feelings-based comment is that I don't like the proposal because I view it _basically_ as "we are killing this feature for the next 3-5 years at a minimum". I don't foresee anyone stepping up to implement the needed compiler changes. Even once they are implemented, the provider API needs to change, and then we are back to square one of waiting a year or two for potential stabilization.

**Comment by jcrevier** [comment]
Hello! Haven't been apart of this discussion so far but have been following this RFC and implementation for a while. 

I like the idea of having some sort of static side casting capability in rust. More reflection like capabilities are something I think rust would benefit from. But as a matter of process, it seems like an entirely different feature than the provider API that requires compiler support and that needs plenty of discussion and probably a separate RFC I would think. 

As a user of Rust, I'm not particularly concerned about the provider API having this (O*M) behavior. I expect values of O and M to usually be very small and really not be a problem in practice. The motivating case for the provider API is errors, which probably aren't on a happy path. If there's  different use case that requires better performance, maybe the provider API isn't what you reach for?

As I see it, the value to be gained by improving the overall ecosystem of errors in Rust is huge and I hope that's not lost on people when considering how to proceed. At every org I've joined that uses rust, I've had to rewrite and refactor the error code to avoid poor error messages, and to get helpful backtraces/ span traces. And it's a always been an uphill battle to get that info from the rust error ecosystem. Holding off on stabilizing the Provider API because of the possibility to do this with a very different fundamental mechanism seems like letting the perfect be the enemy of the good.

**Comment by the8472** [comment]
> Ignoring exact syntax...
> 
>     * Error type `Inner` provides a `Backtrace`
>     * Error type `Outer` wraps an `Inner`
>     * `Outer` can provide a `Backtrace` by delegating to `Inner`
> 
> 
> Notably, `Outer` doesn't statically declare that it will provide a `Backtrace` — it's up to the underlying error. Additionally, it's not based on the _type_ of the underlying error but the _value_. For example, the inner error may be an enum with two variants, one with and one without a backtrace.
> 
> Would your proposed change handle this dynamic behavior?

Wouldn't that be covered by traversing the `Error::cause` hierarchy and then probing the provider for each? That's one of the multipliers I mentioned.


> My more feelings-based comment is that I don't like the proposal because I view it basically as "we are killing this feature for the next 3-5 years at a minimum". 

I'm trying to look at the object-level of API surface and the behavior that follows from it, not the expediency vs. quality meta.

> That being said, making more stuff known statically would change the calculus on how these provided data will be treated from a semver perspective

From a static analysis perspective, yes. But the API still involves a fallible runtime lookup. And I wasn't thinking of making the involved metadata pub.

> I expect values of O and M to usually be very small and really not be a problem in practice.



> The motivating case for the provider API is errors

I would not assume that APIs are only ever useful for their *motivating* use-case and will also end up filling adjacent roles because they're already available as part of the standard library.


> If there's different use case that requires better performance, maybe the provider API isn't what you reach for?

If this were some 3rd party library I wouldn't bat an eye. Putting something in std that amounts to "doesn't work except for small N" into std then this at least warrants more discussion than just assuming it away.

**Comment by the8472** [comment]
Perhaps most of the static knowledge can be achieved with const instead by declaring demand- and provided-sets ahead of time. This way they can be sorted or stuffed into a const hashset. At runtime the provided and demand sets could then be matched to each other in O(max(N, M)) or O(min(N, M)), depending on the chosen data structures.

That doesn't solve the tooling/discovery issue, but at least it would be more scalable.

```rust
#![feature(const_trait_impl)]
#![feature(const_mut_refs)]
use core::marker::PhantomData;


struct DemandFactory<D> {
    demanded: PhantomData<D>
}

impl<D: Default> DemandFactory<D> {
    
    const fn new() -> Self {
        DemandFactory {
            demanded: PhantomData
        }
    }
    
    const fn add_demand<T>(&mut self, store: fn(D, T)) {
        // TODO
    }
}

struct Foo {}

struct MyDemand {
    foo: Option<Foo>
}

impl Default for MyDemand {
    fn default() -> Self { MyDemand {foo: None} }
}

const MY_DEMANDS: DemandFactory<MyDemand> = {
    let mut d = DemandFactory::<MyDemand>::new();
    d.add_demand::<Foo>(|mut my_demand, foo| my_demand.foo = Some(foo));
    d
};
```

**Comment by shepmaster** [comment]
> Wouldn't that be covered by traversing the `Error::cause` hierarchy and then probing the provider for each? That's one of the multipliers I mentioned.

Perhaps I am drastically misreading your initial posts. When you said

> This gets multiplied further when traversing `Error::source()` chains.

I read it as a Very Bad Thing that should not happen and that your proposed solution would thus **disallow**. Am I incorrect? 

You state

> the set of provided properties would have to be statically defined

Given that, I assume the answer to my question ("Would your proposed change handle this dynamic behavior?") is a **no**, but I don't want to put words into your mouth. My reading of your proposal is that we'd have to write code akin to (made-up syntax roughly like yours and extended)

```rust
enum Inner {
    Alpha(Backtrace),
    Beta,
}

impl Inner {
    #[provides]
    const PROVIDES = [Self::whatever_goes_here_for_enum_variants];
}

struct Outer { inner: Inner, status: ExitCode };

impl Outer {
    #[provides]
    const PROVIDES = [
        ...Inner::PROVIDES, // Copy everything from `Inner`... somehow
        Backtrace, // Or maybe we have to duplicate all the types from `Inner`?
        Self::status, // Add our own data
        // How can we choose to supersede `Inner`'s data with our own (or vice-versa)?
    ];
}
```

This give me visions of the annoyances of Java's checked exceptions that grow and grow in the function signature until people give up and say that a function throws any possible exception.

**Comment by the8472** [comment]
> I read it as a Very Bad Thing that should not happen and that your proposed solution would thus disallow. Am I incorrect?

Not quite. I'm saying that error sources provide yet another dimension by which we must parametrize our computational complexity. I.e. an `O` on top of our `M` and `N`. So as various parties add more factors to the whole thing we all must work harder to keep the whole tolerable. E.g. a `O(min(M, N) * O)` is a lot better than `O(M * N * O)`.

I.e. I'm not saying you should do one or the other. I'm just saying that if you also walk the `Error::source` chain then you're increasing the amount of work being done which makes efficiency even more relevant.
I'm just highlighting that people shouldn't only think about the simplest examples but also about more complex software that combines several features. Ideally these features shouldn't compound polynomially.

> (made-up syntax roughly like yours and extended)
>
> ```rust
> struct Outer { inner: Inner, status: ExitCode };
> ```

In this case Inner is not a dynamic type. Not even generic. So with variants 2.) a) and 2.) b) `Outer` could forward the access to `Inner`'s backtrace either with a getter or through a trait implementation.
But yes if you're not walking the `source` chain but still want `Inner`'s provided properties then `Outer` would have to statically redeclare them with my proposal.

**Comment by joshtriplett** [comment]
@the8472 I can definitely understand why in theory that would improve efficiency. And I think it *is* worth avoiding an n^2 operation. However, I think the approach you're proposing implies joining all the things you're asking for together into one call (e.g. asking for "backtrace and exit code" rather than asking for backtrace and then asking for exit code).

1) I think, in practice, many users *won't* be able to join together the things it asks for, because they'll be requested by different parts of code.

2) If users *can* join together the things they ask for, we can already eliminate the quadratic behavior by passing in all the requests in a data structure with constant time lookup and let the provider work with that.

**Comment by joshtriplett** [comment]
(sorry, hit the wrong button)

**Comment by the8472** [comment]
> If users can join together the things they ask for, we can already eliminate the quadratic behavior by passing in all the requests in a data structure with constant time lookup and let the provider work with that.

Well, constructing such a datastructure efficiently isn't trivial. Especially if we want to do it without allocation and in a way that we can send it over the object-safe boundary. I was initially thinking of using sorted arrays (which take NlogN time to construct) and thus doing that at const-eval time to paying that cost every time. And using sorted arrays on one side would yield NlogM runtime. So we'd have to do it on both sides to get down to max(N, M).

If we use hashsets instead then yes, doing it on one side could be sufficient.

> I think, in practice, many users won't be able to join together the things it asks for, because they'll be requested by different parts of code.

In that case it would be better to do it on the provider side to build a provided-set at const-time and then let individual demands go through that.

**Comment by the8472** [comment]
If that can be made to work it would solve the performance issue, but the discoverability-issue would remain. The provided types won't show up in rustdoc or rust-analyzer.

That is of course already true for `dyn Any`, but making it part of a central type like `Error` would increase the impact.

If a type only provides fields the provided-set could be created by a derive and annotations which could then be documented.
E.g. 

```rust
#[derive(Provider)]
struct FooError {
   #[provide]
   trace: Backtrace
}
```

**Comment by Amanieu** [comment]
We discussed this again in the libs meeting today, and feel more confident that this API can eventually be stabilized. However we would like the following changes to be made:

- The `Provider` trait should be deleted and its functionality should be merged into `Error`. We are happy to only provide an API that is only usable with `Error`. If there is demand for other uses then this can be provided through an external crate.

- We are happy to have both `request_val` and `request_ref`. 

- We feel that the names "demand" and "request" are somewhat synonymous and would like only one of those to be used for better consistency.

- We took a look at the generated code and found that LLVM is unable to optimize multiple `.provide_*` calls into a switch table because each call fetches the type id from `Erased::type_id` separately each time and the compiler doesn't know that these calls all return the same value. This should be fixed.

**Comment by noamtashma** [comment]
Users should always wrap the types they request in newtypes in order for the meaning of the value to be clear.

Maybe add a marker trait "DynRequestable" (not the best name ofc) which implies that this type was indeed meant to be used with this interface, within a specific context?

That will enforce that this interface can't be used with possibly confusing common types like `String`.

**Comment by shepmaster** [comment]
> wrap the types they request in newtypes in order for the meaning of the value to be clear.

I agree with this — a `struct Name(String)` is usually better than just a `String` and that applied beyond the provider API.

> enforce that this interface can't be used with possibly confusing common types

I don't agree with this — who gets to decide what is "common"? When there's concrete ambiguity, people can wrap in a newtype and that should be enough.

**Comment by noamtashma** [comment]
To say the same thing in another way:
If you introduce a marker trait like "DynRequestable", then it will enforce that there is an agreed upon meaning to this type.

If this type was not meant to be used with the provider api, it will not implement "DynTequestable", which will force the user to wrap it in and implement it themselves.

**Comment by shepmaster** [comment]
You are talking about implementation details but I am talking about whether or not the idea is good in the first place. I do not think it is worthwhile to artificially limit what types are available in the provider API.

The enforcement is trivially broken:

```rust
struct Provided<T>(pub T);
impl<T> DynRequestable for Provided<T> {}
```

Even if a type has the marker trait, that doesn't mean that it's meaningful *in a specific context*. Maybe in my usage providing `ExitCode` is reasonable, but in another usage that would be ambiguous and should actually be `ServerExit(ExitCode)` and `ClientExit(ExitCode)`. Does that mean that `ExitCode` should not be allowed? What if a third usage means that you need `ApiServerExit(ServerExit)` and `DatabaseServerExit(ServerExit)` — we can't retroactively remove the trait from `ServerExit`.

Can you provide concrete, specific problems that have occurred because of the lack of this marker trait? How would this differ from a hypothetical `HashMapValue` marker trait that is required to put a value into a `HashMap` that we seem to be OK not having?

**Comment by yaahc** [comment]
>     * We took a look at the generated code and found that LLVM is unable to optimize multiple `.provide_*` calls into a switch table because each call fetches the type id from `Erased::type_id` separately each time and the compiler doesn't know that these calls all return the same value. This should be fixed.

I believe @eddyb prototyped a form that optimizes better in one of the other provider related issue/pr threads though it would necessitate some API changes. 

Here's the link if anyone wants to take ownership of resolving this concern (not that this is necessarily the only solution): https://github.com/rust-lang/rfcs/pull/3192#issuecomment-1018020335

**Comment by yaahc** [comment]
Regarding the "request" vs "demand" naming question, I'm gonna go ahead and share my opinion incase it helps make a decision, noting that I don't have any strong objections in either direction.

I'd choose to call it `request` personally, I think mostly because it feels more polite, but also it does happen to be more consistent with the original API surface from [dyno](https://docs.rs/dyno/latest/dyno/provider/struct.Request.html).

**Comment by waynr** [comment]
> Pick either demand or request and use that name consistently.

> Delete the Provider trait and merge its functionality into Error.

@yaahc hey I know I only agreed to work on the RFC for `error_generic_member_access` but to get better familiarity with the code it made sense for me to get started working through some of the unaddressed feedback: https://github.com/rust-lang/rust/pull/113464

I only realized just now that you only asked me to work on the RFC, I hope I didn't jump the gun with that PR!

**Comment by bryanburgers** [comment]
There’s interest in using this pattern in serde: https://github.com/serde-rs/serde/pull/2420

If the Provider trait and `core::any::request_*` drop out, how can non-Error userspace code create a `Demand`/`Request`?

**Comment by waynr** [comment]
> If the Provider trait and core::any::request_* drop out, how can non-Error userspace code create a Demand/Request?

One idea I had while working on my PR was to add a constructor to `Request`, eg
```
  impl<T> Serialize for StringSet<T> where T: Serialize {
      fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
      where
          S: serde::Serializer,
      {
          let request = Request::new_tagged<serde_dynamo::StringSetSerializer>();
          if let Some(string_set_serializer) = serializer.provide(request) {
              self.0.serialize(string_set_serializer)
          } else {
              self.0.serialize(serializer)
          }
      }
  }
```

I'm still fairly new to this stuff but I believe this should be possible with a definition of `Request::new_tagged` that looks something like this:
```
impl Request {
  pub fn new_tagged<'a, I>() -> Self<I:Reified>
  where
    I: tags::Type<'a>,
  {
    let mut tagged = core::any::TaggedOption<'a, I>(None);
    tagged.as_request()
  }
}
```

With this approach, `Serializer` itself would have add the `provide` method, similar to `Error` in my PR.

It's not totally clear to me if that would work for your case, I'm just offering a suggestion that would keep in line with the maintainer's request to leave out the `Provider` trait.

**Comment by bryanburgers** [comment]
> > If the Provider trait and core::any::request_* drop out, how can non-Error userspace code create a Demand/Request?
> 
> One idea I had while working on my PR was to add a constructor to `Request`, eg
> 
> ```
>   impl<T> Serialize for StringSet<T> where T: Serialize {
>       fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
>       where
>           S: serde::Serializer,
>       {
>           let request = Request::new_tagged<serde_dynamo::StringSetSerializer>();
>           if let Some(string_set_serializer) = serializer.provide(request) {
>               self.0.serialize(string_set_serializer)
>           } else {
>               self.0.serialize(serializer)
>           }
>       }
>   }
> ```
> 
> I'm still fairly new to this stuff but I believe this should be possible with a definition of `Request::new_tagged` that looks something like this:
> 
> ```
> impl Request {
>   pub fn new_tagged<'a, I>() -> Self<I:Reified>
>   where
>     I: tags::Type<'a>,
>   {
>     let mut tagged = core::any::TaggedOption<'a, I>(None);
>     tagged.as_request()
>   }
> }
> ```
> 
> With this approach, `Serializer` itself would have add the `provide` method, similar to `Error` in my PR.
> 
> It's not totally clear to me if that would work for your case, I'm just offering a suggestion that would keep in line with the maintainer's request to leave out the `Provider` trait.

Would `tags::Type` need to be added to the public API, then?

**Comment by waynr** [comment]
> Would tags::Type need to be added to the public API, then?

I forgot about that, but I think you're right. The `core::any::tags` module along with the `core::any::TaggedOption` type would probably need to be exposed publicly. This is because the `Request` type itself doesn't provide any methods for accessing the contents of a request, just methods for providing a response to any types that a "provider" (not the trait, just the general term for any type with a method that takes a `Request` for the sake of fulfilling it) is able to respond with.

Kind of strangely, I don't think `Request` is or was ever exposed directly to the user requesting a value/ref (as opposed to the user writing an `Error` implementation). The user requesting a value/ref is meant to use `<dyn Error>.request_value` or `<dyn Error>.request_ref` both of which as a matter of implementation detail end up constructing a `TaggedOption`, turning that into a `Request`, then responding with an `Option<T>`. 

So because `core::any::TaggedOption` and `core::any::tags` are private to the standard library and because my PR removes the `Provider` trait, we can only implement methods like `request_value` and `request_ref` for types in the standard library; prior to my PR the `Provider` trait made it possible for `request_value` and `request_ref` to be implemented for all types that implement `Provider`.

**Comment by waynr** [comment]
@bryanburgers just to follow up, I don't think it's likely in the scope of `provide_any` or `error_generic_member_access` that it will be possible to construct `Request` instances as I mentioned or to use them except in relation to the `core::error::Error` trait. For reference:

* May 2023 libs team meeting (summary of requested changes](https://github.com/rust-lang/rust/issues/96024#issuecomment-1554773172)
* July 9 2023 [confirmation that the libs team meeting only wants this functionality to be available for `Error` types](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/error_generic_member_access/near/373733742)

Although my opinion doesn't mean much since I am still new to contributing here, I think it's better to keep the API impact of features like this small initially.

As for your original question:
>  how can non-Error userspace code create a Demand/Request?

One option you could consider is this [dyno crate](https://docs.rs/dyno/0.1.0/dyno/) which offers a very similar API. Would that work for you? I'm sure it would be easier to be enthusiastic about a standard library `Provider` type than an experimental third-party crate :upside_down_face:

**Comment by waynr** [comment]
Not sure what the proper etiquette is for proposing this, but after seeing agreement [from @Amanieu in zulip](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/error_generic_member_access) I would like to propose closing this tracking issue in favor of merging what's left of the Provider API into the [`error_generic_member_access` RFC](https://github.com/rust-lang/rfcs/pull/2895) which I have volunteered to update (I'll be opening a PR against that PR shortly after submitting this comment).

The major motivation behind merging these two issues is that the libs meeting has decided to leave out the `Provider` trait. Without that, all `provide_any` provides is the `Request` type which for the foreseeable future will only be used in the `Error` trait updates proposed in [the `error_generic_member_access` RFC](https://github.com/rust-lang/rfcs/pull/2895); which anyway already mentions a `Request` type more or less identical to what has been discussed here.

Merging the two issues would involve the following steps:

* [ ] first, make sure there are no major disagreements here in this tracking issue (the purpose of my current message)
* [ ] move all outstanding feedback items and next steps from this tracking issue into the [`error_generic_member_access` tracking issue](https://github.com/rust-lang/rust/issues/99301)
* [ ] submit PR to the rfcs repo to remove https://github.com/rust-lang/rfcs/blob/master/text/3192-dyno.md
* [ ] close this issue tracking issue
* [ ] incorporate the `Request` (aka `Demand`) specific elements of RFC 3192 into the WIP RFC 2895

Anyone have any objections to this? I kind of feel awkward jumping in and suggesting it as a newcomer, but again I've had some supportive responses in zulip. I'm happy to take on all this work for the near future as I am currently unemployed anyway.

**Comment by Amanieu** [comment]
> Anyone have any objections to this? I kind of feel awkward jumping in and suggesting it as a newcomer, but again I've had some supportive responses in zulip. I'm happy to take on all this work for the near future as I am currently unemployed anyway.

No objections, this sounds like a good plan. Let me know if you need help with anything.

**Comment by waynr** [comment]
@Amanieu okay cool, sounds like we've given folks long enough to object. I can't do much keyboard work at the moment because i recently smashed one of my fingers (by mistake of course :upside_down_face:) and it needs a few more days to a week to heal before I'm back to my full typing speed. But for now I'll try to assign both tracking issues to myself and start making some of the changes that don't require as much keyboard work.

**Comment by waynr** [comment]
@yaahc I think we've captured everything but 

> LLVM is unable to optimize multiple .provide_* calls into a switch table because each call fetches the type id from Erased::type_id separately each time and the compiler doesn't know that these calls all return the same value. This should be fixed. See Add provide_any module to core rfcs#3192 (comment) for one possible solution (although alternative solutions may be possible).

in the TODOs for the [`error_generic_member_access` tracking issue](https://github.com/rust-lang/rust/issues/99301) so we should be able to close this tracking issue, right?

**Comment by jmjoy** [comment]
How about add an optional key? Just like:

```rust
pub fn provide_value<T>(&mut self, key: Option<&str>, value: T) -> &mut Demand<'a>;
```

then we can provide multi value with the same type, such as:

```rust
struct Sth {
    foo: String,
    bar: String,
}

impl Provider for Sth {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
        demand
            .provide_value::<String>(Some("foo"), self.foo)
            .provide_value::<String>(Some("bar"), self.bar);
    }
}

```

**Comment by waynr** [comment]
FEI, the Provider API has been rejected by the libs meeting folks. For more background, please see my earlier comment in this issue: https://github.com/rust-lang/rust/issues/96024#issuecomment-1629794600

The TL;DR is that the `Provider` trait has been rejected, leaving just the `Request` type (aka the type formerly known as `Demand`) and the new `provide` method on the `Error` trait. The `Request` type and its associated internal-only machinery has been moved from the `core::any` module into `core::error`. This is the state of affairs in latest nightly after #113464 was merged earlier this week.

**Comment by TennyZhuang** [comment]
Is there any third-party crate to maintain the `Provider` trait in the future? Although it may not be suitable to be merged in the `std::any`, it's still a useful pattern in some scenarios.

**Comment by waynr** [comment]
@TennyZhuang the closest thing I know of was mentioned in the notes for this feature's RFC, which can be found here: `[provide-any](https://github.com/nrc/provide-any)`

I don't know how similar this is to what was unstable here in the standard library and there are no published packages.

**Comment by conradludgate** [comment]
There is also the supplier crate that I made in response to this feature being removed. I also consider it unstable but it's a direct lift and shift from the latest provider impl that std had in nightly
